src/dms/dmsScanner.cpp

```diff
@@ -0,0 +1,346 @@
+         }
+         goto error ;
+      }
+
+   done:
+      return rc ;
+   error:
+      recordID.reset() ;
+      _curRID._offset = DMS_INVALID_OFFSET ;
+      goto done ; 
+   }
+
+   void _dmsExtScannerBase::stop()
+   {
+      _next = DMS_INVALID_OFFSET ;
+      _curRID._offset = DMS_INVALID_OFFSET ;
+   }
+
+   _dmsExtScanner::_dmsExtScanner( dmsBlockData *su,
+                                   _dmsMBContext *context,
+                                   dmsExtentID curExtentID,
+                                   DMS_ACCESS_TYPE accessType,
+                                   INT64 maxRecords,
+                                   INT64 skipNum,
+                                   INT32 flag )
+   : _dmsExtScannerBase( su, context, curExtentID, accessType,
+                         maxRecords, skipNum, flag )
+   {
+   }
+
+   _dmsExtScanner::~_dmsExtScanner()
+   {
+   }
+
+   INT32 _dmsExtScanner::_firstInit( pmdEDUCB *cb )
+   {
+      INT32 rc          = SDB_OK ;
+
+      _extRW = _pSu->extent2RW( _curRID._extent, _context->mbID() ) ;
+      _extRW.setNothrow( TRUE ) ;
+      _extent = _extRW.readPtr<dmsExtent>() ;
+      if ( NULL == _extent )
+      {
+         rc = SDB_INVALIDARG ;
+         goto error ;
+      }
+/*      if ( cb->isInterrupted() )
+      {
+         rc = SDB_APP_INTERRUPT ;
+         goto error ;
+      }
+*/
+      if ( !_context->isMBLock( _mbLockType ) )
+      {
+         rc = _context->mbLock( _mbLockType ) ;
+         PD_RC_CHECK( rc, PDERROR, "dms mb lock failed, rc: %d", rc ) ;
+      }
+      if ( !_extent->validate( _context->mbID() ) )
+      {
+         rc = SDB_SYS ;
+         goto error ;
+      }
+
+      _cb   = cb ;
+      _next = _extent->_firstRecordOffset ;
+
+      // unset first run
+      _firstRun = FALSE ;
+
+   done:
+      return rc ;
+   error:
+      goto done ;
+   }
+
+   INT32 _dmsExtScanner::_fetchNext( dmsRecordID &recordID,
+                                     dmsRecordGenerator &generator,
+                                     pmdEDUCB *cb )
+   {
+      INT32 rc                = SDB_OK ;
+      ossValuePtr recordDataPtr ;
+      dmsRecordData recordData ;
+
+      while ( DMS_INVALID_OFFSET != _next && 0 != _maxRecords )
+      {
+         _curRID._offset = _next ;
+         _recordRW = _pSu->record2RW( _curRID, _context->mbID() ) ;
+         _curRecordPtr = _recordRW.readPtr( 0 ) ;
+         _next = _curRecordPtr->getNextOffset() ; 
+
+         SDB_ASSERT( !_curRecordPtr->isDeleted(), "record can't be deleted" ) ;
+
+         recordID = _curRID ;
+         rc = _pSu->extractData( _context, _recordRW, recordData ) ;
+         if ( rc )
+         {
+            PD_LOG( PDERROR, "Extract record data failed, rc: %d", rc ) ;
+            goto error ;
+         }
+         recordDataPtr = ( ossValuePtr )recordData.data() ;
+         generator.setDataPtr( recordDataPtr ) ;
+
+         try
+         {
+            BSONObj obj( recordData.data() ) ;
```

Comment:
why do you reset the data here?

Reply:
it set the bson object in the generator and increment the number of data inside the generator to indicate there is data in the generator